source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,282 @@
+  }
+  return true;
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    // TODO(gsagula): do this lookup without allocating a new string.
+    std::string value{StringViewUtil::cropRight(content_type->value().c_str(), ";")};
+    return config_->contentTypeValues().find(value) != config_->contentTypeValues().end();
+  }
+  return true;
+}
+
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* etag = headers.Etag();
+  if (config_->disableOnEtag()) {
+    return !etag;
+  }
+  if (etag) {
+    absl::string_view value(etag->value().c_str());
+    if (!(value.length() > 2 && (value[0] == 'w' || value[0] == 'W') && value[1] == '/')) {
+      headers.removeEtag();
+    }
+  }
+  return true;
+}
+
+bool GzipFilter::isLastModifiedAllowed(HeaderMap& headers) const {
+  return !(config_->disableOnLastModified() && headers.LastModified());
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length;
+    return StringUtil::atoul(content_length->value().c_str(), length)
+               ? length >= config_->minimumLength()
+               : false;
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding &&
+          StringViewUtil::findToken(transfer_encoding->value().c_str(), ",",
+                                    Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isTransferEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  if (transfer_encoding) {
+    for (auto header_value : StringViewUtil::splitToken(transfer_encoding->value().c_str(), ",")) {
+      if (header_value == Http::Headers::get().TransferEncodingValues.Gzip ||
+          header_value == Http::Headers::get().TransferEncodingValues.Deflate) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+void GzipFilter::insertVary(HeaderMap& headers) {
+  const Http::HeaderEntry* vary = headers.Vary();
+  if (vary) {
+    auto header_values = StringViewUtil::splitToken(vary->value().c_str(), ",");
+    bool has_accept_encoding{false};
+
+    std::ostringstream oss;
+    for (auto it = header_values.begin(); it != header_values.end(); it++) {
+      if (*it != Http::Headers::get().VaryValues.Wildcard) {
+        oss << *it;
+      }
+      if (*it != header_values.back() && it != header_values.begin()) {
+        oss << ", ";
+      }
+      if (*it == Http::Headers::get().VaryValues.AcceptEncoding) {
+        has_accept_encoding = true;
+      }
+    }
+
+    if (!has_accept_encoding) {
+      oss << ", " << Http::Headers::get().VaryValues.AcceptEncoding;
```

Comment:
This seems more complex than necessary.  Why not just do
  if (!findToken(vary_string, ",", "Accept-Encoding")) {
      headers.insertVary().value(vary_string + ", Accept-Encoding");

Reply:
👍 💯

Reply:
Revisiting the current design, I realized that the above wouldn't work. If I interpreted the RFC correctly, the filter would have to also look up for '*' value and remove it before sending it to the client:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary
```
*
Each request for a URL is supposed to be treated as a unique and uncacheable request. A better way to indicate this is to use Cache-Control: private, which is clearer to read and also signals that the object shouldn't be stored ever.
``` 

https://tools.ietf.org/html/rfc7231#page-70
```
A Vary field value of "*" signals that anything about the request
might play a role in selecting the response representation, possibly
including elements outside the message syntax (e.g., the client's
network address).  A recipient will not be able to determine whether
this response is appropriate for a later request without forwarding
the request to the origin server.  A proxy MUST NOT generate a Vary
field with a "*" value.
```

Reply:
IMO if you have Vary: *, you shouldn't remove it, and you don't need to add Accept-Encoding either. That quote "a proxy MUST NOT generate Vary:*" is confusing.  I think maybe generating is not good, but transmitting is OK? I will consult an expert & get back to you.

Reply:
That would be super helpful. I research a bit more about it too. Thanks!

Reply:
Got one response from a well-known standards guy (Ilya Grigorik) but not yet the Source Of Truth (Mark Nottingham):

From:me:

> Hi Mark & Ilya, sorry to bother you; I have what I think is a quick question.
> 
> I came across a section of the http spec whose wording inspires me to seek clarification.
> 
> https://tools.ietf.org/html/rfc7231#page-70
> 
>    A Vary field value of "*" signals that anything  about the request
>    might play a role in selecting the response representation, possibly
>    including elements outside the message syntax (e.g., the client's
>    network address).  A recipient will not be able to determine whether
>    this response is appropriate for a later request without forwarding
>    the request to the origin server.  *A proxy MUST NOT generate a Vary
>    field with a "*" value.*
> 
> (boldface added by me)
> 
> I don't think this is suggesting that proxies strip Vary:* sent from origin; just that they not generate it themselves.  Correct?

From Ilya:
> That's my read too, but I'll defer to Mark.